RAG Tutorial Notebook

mihaela-bornea opened the pull request:
# Pull Request      

## What does this PR do?

RAG tutorial notebook and changes to simplify the RAG code.       

## Description

We include a reader factory to create generative readers with HuggingFace models, including the ones we train with PrimeQA, and OpenAI models. 
Added a RAG pipeline that includes the retriever as a searchable corpus and the Generative Reader.


### Checklist

Review the following and mark as completed:
- [ ] [Tag an issue or issues](#what-does-this-pr-do) this PR addresses.
- [ ] Added [description](#description) of changes proposed.
- [ ] Review requested as [appropriate](#request-review).
- [ ] Version bumped as [appropriate](#Versioning).
- [ ] New classes, methods, and functions documented.
- [ ] Documentation for modified code is updated.
- [ ] Built documentation to confirm it renders as expected (see [here](development.md)).
- [ ] Code cleaned up and commented out code removed.
- [ ] Tests added to ensure all functionalities tested at >= 60% unit test coverage (see [here](development.md)).
- [ ] Code cleaned up and commented out code removed.
- [ ] Release created [as needed](#after-merging) after merging.

vishwajeetkumar93 commented on notebooks/ir/dense/dpr_tutorial.ipynb:
```diff
@@ -63,12 +69,113 @@
    "source": [
     "# Please update to point to your collection tsv format id\\ttext\\ttitle with these headers\n",
-    "path_to_collection_tsv_file=\"../path_to_tsv/sample.tsv\"\n",
+    "path_to_collection_tsv_file=\"../../../../ibm-generative-ai-cookbooks/notebooks/docs_and_qs/psgs.tsv\"\n",
```
Is this path hardcoded?

vishwajeetkumar93 commented on primeqa/ir/scripts/test_SC.py:
```diff
@@ -31,6 +31,6 @@ def select_column(data, col):
                     out.write("\t".join([queries[q]['id'], ans, str(rank+1), str(score)])+"\n")
     answers = []
-
+ 
```
Please remove empty lines.

vishwajeetkumar93 commented on primeqa/pipelines/qa_pipeline.py:
```diff
@@ -34,3 +35,29 @@ def run(self, input_texts: List[str], prefix="", suffix="", use_retriever=True):
             result[i] = i_result
 
         return result
+
```
Please remove empty lines.

vishwajeetkumar93 reviewed (approved):
LGTM. I have left some comments.

bhavani105 commented on primeqa/components/reader/reader_factory.py:
```diff
@@ -0,0 +1,98 @@
+                                    contexts=contexts,
+                                    example_ids=example_ids,
+                                    *args,
+                                    *kwargs)
```
Please add doctoring for the class and argument

bhavani105 commented on primeqa/services/factories.py:
```diff
@@ -28,8 +28,8 @@
     ExtractiveReader.__name__: ExtractiveReader,
-    GenerativeReader.__name__: GenerativeFiDReader,
-    PromptReader.__name__: PromptFLANT5Reader,
+    GenerativeBaseReader.__name__: GenerativeFiDReader,
```
Should this be  GenerativeFidReader.__name__

bhavani105 reviewed (commented):
Looks good.
Left some comments

bhavani105 commented on docs/tutorials/02_RAG_Generator.ipynb:
```diff
@@ -0,0 +1,257 @@
+    "%%bash\n",
+    "\n",
+    "pip install --upgrade pip\n",
+    "pip install primeqa"
```
pip install --upgrade primeqa

mihaela-bornea commented on tests/primeqa/services/rest_server/test_readers.py:
```diff
@@ -25,6 +25,6 @@ def test_get_readers(mock_client):
         "ExtractiveReader",
-        "GenerativeReader",
-        "PromptReader",
+        "GenerativeBaseReader",
```
I think I will leave the base class in there for now.

mihaela-bornea commented:
I addressed all comments except:
1. Still using the GenerativeBaseReader for now since we might add another generative reader, in addition to FiD.
2. I did not update the RAG notebook since has been previously reviewed

bhavani105 reviewed (approved):
LGTM
